src/crimson/net/Socket.h

```diff
@@ -178,7 +178,7 @@ class FixedCPUServerSocket
       lo.set_fixed_cpu(ss.cpu);
       ss.listener = seastar::listen(s_addr, lo);
     }).handle_exception_type([addr] (const std::system_error& e) {
-      if (e.code() == error::address_in_use) {
+      if (e.code() == std::errc::address_in_use) {
```

Comment:
strictly speaking, this is not right. we cannot assume that `system_category().default_error_condition(address_in_use)` returns what `generic_category()` returns. please see https://github.com/llvm-mirror/libcxx/blob/7c3769df62c0b3820130aa868397a80a042e0232/src/system_error.cpp#L197-L204 and https://github.com/gcc-mirror/gcc/blob/d3f09bd8151666f5a180fc38337ea835e109caaf/libstdc%2B%2B-v3/src/c%2B%2B11/system_error.cc#L70-L321 for more details.

Reply:
> we cannot assume that system_category().default_error_condition(address_in_use) returns what generic_category() returns.

Maybe the simplest way is to add static_assert to make sure our assumptions are correct in the environments we want to support. It current works in my environment and I guess it will mostly work in linux, see https://github.com/gcc-mirror/gcc/blob/master/libstdc++-v3/config/os/generic/error_constants.h and https://github.com/llvm-mirror/libcxx/blob/master/src/include/config_elast.h.

In case those assertions are failed in some of our target environments, it is still not reasonable to define our errc counterparts inside crimson::net::errors:
- We will have 3 different sources of error_code for "`invalid_argument` and the like" which are really confusing: 1) customly defined in crimson/net/Errors.h, 2) used by seastar with system category, and 3) used by errorator (`crimson::ct_error::invarg`) and stdlib with generic category.
- We will need to remember to always compare `invalid_argument ...` with `crimson::net::error::invalid_argument ...`, which also means those errors should be splitted out to a more common directory and namespace like `src/crimson/common`.

Reply:
> > we cannot assume that system_category().default_error_condition(address_in_use) returns what generic_category() returns.
> 
> Maybe the simplest way is to add static_assert to make sure our assumptions are correct in the environments we want to support. It current works in my environment and I guess it will mostly work in linux, see https://github.com/gcc-mirror/gcc/blob/master/libstdc++-v3/config/os/generic/error_constants.h and https://github.com/llvm-mirror/libcxx/blob/master/src/include/config_elast.h.
> 
> In case those assertions are failed in some of our target environments, it is still not reasonable to define our errc counterparts inside crimson::net::errors:
> 
> * We will have 3 different sources of error_code for "`invalid_argument` and the like" which are really confusing: 1) customly defined in crimson/net/Errors.h, 2) used by seastar with system category, and 3) used by errorator (`crimson::ct_error::invarg`) and stdlib with generic category.

it's confusing. but IMHO, not *that* confusing, i think that's exactly what `std::error_category` is for, if it's an error of `system_category`, then the error is caused by invalid arguments passed to system call, and the error should be mapped to a POSIX errno . if it's an error of `generic_category`, then the error code cannot be mapped to a POSIX errno. the invalid_argument from our category is, well, not caused by system call or underlying libc calls.

> * We will need to remember to always compare `invalid_argument ...` with `crimson::net::error::invalid_argument ...`, which also means those errors should be splitted out to a more common directory and namespace like `src/crimson/common`.